wordpress-mobile/WordPress-Android
# Issue/9728 posts remote preview

### Fixes #9728 
This PR adds support for Remote Preview of blog Posts vs the currently available local preview.
To this goal we need that the Post the user wants to preview is previously synchronized with the server.
In this PR we are mainly interested in the following categories of Posts:

1. Draft: this exists already on the server actually
2. Local Draft: a draft that exists on only the user's device (both entirely or partially)
3. New Post
4. Published and Scheduled Post with local changes

All of these will be saved locally and synched with the server if needed. To avoid the unintended publication of a published post for the category 4 the remote auto save will be used instead.
(**NOTE**: this PR needs the FluxC modifications for remote auto save available in the _master-remote-auto-save_ branch, see 18205856d74bb04747484182a056741275d97efb).

### User Flow
1. User asks for preview from Posts List or from Editor
2. In case of draft with no modification the remote preview is directly opened
3. In case of new post or local draft, the user is presented with the following progress dialog while the post is uploading and then the remote preview is shown
<p align="center">
  <img width="300" src="https://user-images.githubusercontent.com/47797566/59987617-1de08100-963b-11e9-954c-672173654c1f.png">
</p>
4. In case of published or scheduled post with local changes, the user is presented with the following dialog while the post is remote auto save and then the remote preview is shown
<p align="center">
  <img width="300" src="https://user-images.githubusercontent.com/47797566/59988484-7dd72780-963b-11e9-84be-b89aff91764d.png">
</p>
Special cases:

1. If there is no network, the following view is shown
<p align="center">
  <img width="300" src="https://user-images.githubusercontent.com/47797566/60483312-37b24180-9c95-11e9-981d-48c75233a477.png">
</p>
2. If the connected site is a self-hosted NOT Jetpack powered site, the following is shown (see  https://href.li/?https://github.com/wordpress-mobile/WordPress-iOS/pull/11944#pullrequestreview-251304944 .
<p align="center">
  <img width="300" src="https://user-images.githubusercontent.com/47797566/60483354-59abc400-9c95-11e9-8e7d-bdcecb75e599.png">
</p>

3. If an error occurs during preview operation the following snackbar is shown
<p align="center">
  <img width="300" src="https://user-images.githubusercontent.com/47797566/60483599-6aa90500-9c96-11e9-9551-b0d4d23d9fea.png">
</p>


### Background Implementation notes
- In the Editor and in the Posts List, a state machine was added to manage the state of the preview and the presented progress dialog. 
- An EventBus sticky event `PostPreviewingInEditor` was defined in `EditPostActivity` to signal when the preview is ongoing from the editor. This way we can make some event filtering in the Posts List.
- Since all the upload operations are currently expecting an `OnPostUploaded` event, when we are remote auto saving, we chain on remote auto save response receiving a OnPostUploaded event to conclude the uploading transactions. **NOTE:** in case of errors during uploading, the UI currently present to the user the possibility to Retry the upload. For published posts that fail during remote auto saving for preview, as first solution (but in my opinion it can make sense) we skip the error condition (and an eventual retry that can publish the local changes) by not relaunching the error condition in the OnPostUploaded. In all other cases the user is as usual triggered to Retry upload.

### To test:
1. To be able to ask for preview in the posts list, you need to create local changes; this can be easily done in the editor, and published posts simply keep them in the posts list if you don't explicitly publish them. For drafts you can exit from the editor while in offline mode.
2. To get a failure during upload for preview I used the following: in a self hosted with Jetpack site, disable the jetpack plugin and ask for a preview. You will get an error during preview operations. (I'm sure there are better solutions, but it worked for me 🙂).

This said:

**Test Case 1**
1. Ask for preview connecting to a self-hosted not jetpack site and observe you get the Preview Unavailable view.

**Test Case 2**
1. Connected to a WPcom site, put device offline and ask for preview. Observe you get the no network webview.

**Test Case 3**
1. From the editor, ask for preview on draft with no local changes
2. No progress shown and you get the remote preview

**Test Case 4**
1. Ask for preview on draft with local changes or new post
2. Observe the Saving progress and then get the remote preview

**Test Case 5**
1. Ask for preview on published post with local changes
2. Observe the Generating Preview progress and then get the remote preview
3. repeat previous steps for scheduled post with local changes

**Test Case 6**
1. Create a test scenario for failure during preview creation (for example with the disable of the jetpack plugin as said above)
2. Ask for preview for a draft with local changes and observe you get the following error snackbar 
<p align="center">
  <img width="300" src="https://user-images.githubusercontent.com/47797566/60483698-e30fc600-9c96-11e9-90ec-75aafe0aa561.png">
</p>

and you are allowed to Retry upload
<p align="center">
  <img width="300" src="https://user-images.githubusercontent.com/47797566/59990448-7ddb2580-9643-11e9-8709-bd1549ddfe8e.png">
</p>

**Test Case 7**
1. Create a test scenario for failure during preview creation (for example with the disable of the jetpack plugin as said above)
2. Ask for preview for a published post with local changes and observe you get the same error snackbar as before but you are not presented to Retry
<p align="center">
  <img width="300" src="https://user-images.githubusercontent.com/47797566/59990585-f5a95000-9643-11e9-80de-61c30cbd0384.png">
</p>
3. repeat previous steps for scheduled post with local changes

### Open Items
These are open items that will be faced with dedicated Issue/PR and probably need further discussion. I opened related Issues for them and will follow up based on how things progress here.
- [x] Scheduled posts should be assimilated to published posts and treated the same; currently FluxC allows remote auto save of only Published posts. So need to modify this behavior in FluxC. This current PR implementation tries to manage them as draft. (see [FluxC#1297](https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/1297) )
- [ ] Trashed posts management in preview. (see #10106)
- [x] Menu items and actions available in the WPWebViewActivity when used to preview posts in each of the above status. (see #10107)
- [ ] There is a [known issue](https://github.com/wordpress-mobile/WordPress-iOS/issues/11684) for Jetpack powered sites and remote auto save; I think there is already API side work triggered from that so will follow it for resolution.
- [ ] Remote preview operation and dialog cancellation: see #10141. 
 
**Update release notes:**

- [x] Since there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.
